### PR TITLE
Add CIQ variational strategy

### PR DESCRIPTION
**Pull request by gpleiss**

### CODE REVIEW

File: gpytorch/variational/ciq_variational_strategy.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        """
+        Compute the KL divergence between the variational inducing distribution :math:`q(\mathbf u)`
+        and the prior inducing distribution :math:`p(\mathbf u)`.
+
+        :rtype: torch.Tensor
+        """
+        if hasattr(self, "_memoize_cache") and "kl" in self._memoize_cache:
+            return self._memoize_cache["kl"]
+        else:
+            raise RuntimeError
```
Comment (jacobrgardner) [inline gpytorch/variational/ciq_variational_strategy.py:211]:
```suggestion
            raise RuntimeError("KL divergence for CIQ should be computed during forward calls.")
```

File: gpytorch/variational/ciq_variational_strategy.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        expec_mat_grad = torch.add(
+            (interp_var_grad * interp_term) @ interp_term.transpose(-1, -2),
+            (kl_div_grad.unsqueeze(-1).unsqueeze(-1) * (eye - prec).mul(0.5)),
+        )
+
+        # We're done!
+        return interp_term_grad, expec_vec_grad, expec_mat_grad, None  # Extra "None" for the kwarg
+
+
+class VD(Module):
```
Comment (jacobrgardner) [inline gpytorch/variational/ciq_variational_strategy.py:118]: `_CIQNaturalVariationalDistribution`?

File: gpytorch/variational/ciq_variational_strategy.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    def prior_distribution(self):
+        zeros = torch.zeros_like(self._variational_distribution.natural_vec)
+        ones = torch.ones_like(zeros)
+        res = MultivariateNormal(zeros, DiagLazyTensor(ones))
+        return res
+
+    @property
+    @cached(name="variational_distribution_memo")
+    def variational_distribution(self):
+        raise RuntimeError
```
Comment (jacobrgardner) [inline gpytorch/variational/ciq_variational_strategy.py:159]:
```suggestion
        raise RuntimeError("CIQVariationalStrategy uses a built in variational distribution")
```

File: gpytorch/variational/ciq_variational_strategy.py (hunk 2 above)
Comment (gpleiss) [reply gpytorch/variational/ciq_variational_strategy.py:118]: I'm changing up CIQVariationalStrategy to accept any type of variational distribution - so it should be called with the standard NaturalVariationalDistribution instead.

**Review by jacobrgardner — APPROVED** [review]
This all looks good to me now

**Pull request merged**